server/src/game-implementations/games/ddr.ts

```diff
@@ -0,0 +1,332 @@
+	percent: GoalFmtPercent,
+
+	// don't want commas
+	score: (v) => `Get a score of ${v} on`,
+};
+
+const DDR_GOAL_OO_FMT: GPTGoalFormatters<"ddr:DP" | "ddr:SP"> = {
+	percent: GoalOutOfFmtPercent,
+	// don't insert commas or anything.
+	score: (m) => m.toString(),
+};
+
+const DDR_GOAL_PG_FMT: GPTGoalProgressFormatters<"ddr:DP" | "ddr:SP"> = {
+	percent: (pb) => `${pb.scoreData.percent.toFixed(2)}%`,
+
+	// 4519 -> "4519". Don't add commas or anything.
+	score: (pb) => pb.scoreData.score.toString(),
+
+	lamp: (pb) => {
+		return pb.scoreData.lamp;
+	},
+	grade: (pb, gradeIndex) =>
+		GradeGoalFormatter(
+			DDR_GBOUNDARIES,
+			pb.scoreData.grade,
+			pb.scoreData.score,
+			DDR_GBOUNDARIES[gradeIndex]!.name,
+			(deltaPercent) => {
+				return Math.round(deltaPercent / 1_000_000).toFixed(2);
+			}
+		),
+};
+
+export const DDR_SCORE_VALIDATORS: Array<ScoreValidator<"ddr:DP" | "ddr:SP">> = [
+	(s) => {
+		const { MARVELOUS, PERFECT, GREAT, GOOD, OK, MISS } = s.scoreData.judgements;
+
+		if (
+			IsNullish(MARVELOUS) ||
+			IsNullish(PERFECT) ||
+			IsNullish(GREAT) ||
+			IsNullish(GOOD) ||
+			IsNullish(OK) ||
+			IsNullish(MISS)
+		) {
+			return;
+		}
+
+		const stepScore = 1_000_000 / (MARVELOUS + PERFECT + GREAT + GOOD + OK + MISS);
+		const calculatedScore =
+			Math.floor(
+				(stepScore * (MARVELOUS + OK) +
+					(stepScore - 10) * PERFECT +
+					((stepScore * 3) / 5 - 10) * GREAT +
+					(stepScore / 5 - 10) * GOOD) /
+					10
+			) * 10;
+
+		if (calculatedScore !== s.scoreData.score) {
+			return `Expected calculated score from judgements of ${calculatedScore} to equal score of ${s.scoreData.score}.`;
+		}
+	},
+	(s) => {
+		const { MARVELOUS, PERFECT, GREAT, GOOD, MISS } = s.scoreData.judgements;
+
+		if (
+			IsNullish(MARVELOUS) ||
+			IsNullish(PERFECT) ||
+			IsNullish(GREAT) ||
+			IsNullish(GOOD) ||
+			IsNullish(MISS)
+		) {
+			return;
+		}
+
+		switch (s.scoreData.lamp) {
+			case "FULL COMBO": {
+				if (MISS > 0) {
+					return `Cannot have a FULL COMBO with more than 0 MISS`;
+				}
+
+				break;
+			}
+
+			case "GREAT FULL COMBO": {
+				if (MISS > 0 || GOOD > 0) {
+					return `Cannot have a GREAT FULL COMBO with more than 0 MISS and GOOD`;
+				}
+
+				break;
+			}
+
+			case "PERFECT FULL COMBO": {
+				if (MISS > 0 || GOOD > 0 || GREAT > 0) {
+					return `Cannot have a PERFECT FULL COMBO with more than 0 MISS, GOOD and GREAT`;
+				}
+
+				break;
+			}
+
+			case "MARVELOUS FULL COMBO": {
+				if (MISS > 0 || GOOD > 0 || GREAT > 0 || PERFECT > 0) {
+					return `Cannot have a MARVELOUS FULL COMBO with anyhing else than MARVELOUS judgements`;
+				}
+
+				break;
+			}
+
+			default:
+		}
+	},
+];
+
+export const DDR_IMPL: GPTServerImplementation<"ddr:DP" | "ddr:SP"> = {
+	chartSpecificValidators: {},
+	classDerivers: {
+		flare: (ratings) => {
+			const flarePoints = ratings.flareSkill;
+
+			if (IsNullish(flarePoints)) {
+				return null;
+			}
+
+			switch (true) {
```

Comment:
switch true makes me feel a pit in my stomach but it's not really your fault that this language sucks.

I'll clean this up in my own time, since the code you've written is fine And Works.